smart/mesh_tools.py

```diff
@@ -1408,9 +1415,20 @@ def load_mesh(
     dim = mesh.geometric_dimension()
     mf_cell = d.MeshFunction("size_t", mesh, dim)
     mf_facet = d.MeshFunction("size_t", mesh, dim - 1)
+    subdomains = []
 
     with d.HDF5File(mesh.mpi_comm(), str(filename.with_suffix(".h5")), "r") as hdf5:
         hdf5.read(mf_cell, f"/mf{dim}")
         hdf5.read(mf_facet, f"/mf{dim-1}")
-
-    return LoadedMesh(mesh=mesh, mf_cell=mf_cell, mf_facet=mf_facet, filename=filename)
+        h = h5py.File(str(filename.with_suffix(".h5")), "r")
+        all_keys = list(h.keys())
+        if len(all_keys) > 3:
+            for key in all_keys[3:]:
+                cur_dim = int(key[-1])
+                mf_cur = d.MeshFunction("size_t", mesh, cur_dim)
+                hdf5.read(mf_cur, f"/{key}")
+                subdomains.append(mf_cur)
```

Comment:
This looks a little bit dangerous. First of all, you are opening the same file with `h5py` that is already open with `dolfin.HDFFile`. Second, here we assume that mesh functions are saved in a specific order and that what is saved is a MeshFunction. 

Also, bringing in a new dependency (`h5py`) which is know to be tricky to install.

I am wondering if is is better if the user can provide these keys as an argument when loading the mesh, e.g

```python
def load_mesh(
    filename: Union[pathlib.Path, str],
    comm: MPI.Intracomm = MPI.COMM_WORLD,
    mesh: Optional[d.Mesh] = None,
    extra_keys: Optional[List[str]] = None,
) -> LoadedMesh:
    if extra_keys is None:
        extra_keys = []
```
and then just reading the meshfunctions from those keys.

Reply:
Yes, this would be much better, I'll make this change today, thanks!